Add CssClass property to ValidationSummary

ChristopherHaws opened the pull request:
# Add CssClass property to ValidationSummary
API Proposal: #46461

## Description
Add `CssClass` property to `ValidationSummary` and append the contents to the resulting `class` attribute.

Fixes #43860, Closes: #46461

ChristopherHaws committed:
Add CssClass property to ValidationSummary

ChristopherHaws commented on src/Components/Web/test/Forms/ValidationSummaryTest.cs:
```diff
@@ -0,0 +1,115 @@
+        var ulFrameIndex = Array.IndexOf(batch.ReferenceFrames, ulFrame);
+        var ulFrameSubtree = batch.ReferenceFrames.Skip(ulFrameIndex).Take(ulFrame.ElementSubtreeLength);
+
+        var ulClassFrame = ulFrameSubtree.FirstOrDefault(x =>
```
I'm not used to writing tests for components without bUnit. Is this the easiest way to find the resulting `class` attribute for the generated `ul` element?

SteveSandersonMS replied:
Typically we don't write unit tests for this kind of component rendering. Instead we write end-to-end (E2E) tests that rendering into a real browser and are automated with Selenium.

If you have a look at `src\Components\test\E2ETest\Tests\FormsTest.cs` and `TypicalValidationComponent.razor` you'll see some examples. It should be reasonably straightforward to add a further case following this pattern to show that `ValidationSummary` renders a supplied class name. These E2E tests can be executed as normal tests via `dotnet test` or the VS test window, but if you run the whole suite of them it can take a while, like 15 minutes or more.

ChristopherHaws replied:
Sounds good. I switched this to an E2E test. Had a hell of a time getting selenium to run because the web driver requires v103 of chrome but the latest version of chrome is v105 and there is no way to download an old version of chrome. Ended up having to find a build of chromium for v103. You may want to add this info to the build docs. I remember now why I switched away from selenium and started using playwright instead :)

ChristopherHaws commented on src/Components/Web/src/Forms/ValidationSummary.cs:
```diff
@@ -73,7 +78,7 @@ protected override void BuildRenderTree(RenderTreeBuilder builder)
 
                 builder.OpenElement(0, "ul");
-                builder.AddAttribute(1, "class", "validation-errors");
+                builder.AddAttribute(1, "class", string.IsNullOrWhiteSpace(CssClass) ? "validation-errors" : $"validation-errors {CssClass}");
```
As I state [here](https://github.com/dotnet/aspnetcore/issues/43860#issuecomment-1244145778), other components seem to have `CssClass` as a get only property and use `AttributeUtilities.CombineClassNames` in `OnParametersSet` to append the class name passed in via `AdditionalAttributes`. Let me know if I should update this code to do that as well.

Thanks! :)

SteveSandersonMS replied:
Yes, sorry, you're right. It would be better to align more with the pattern we have elsewhere of extracting the `class` value from `AdditionalAttributes`, with `CssClass` being a readonly property that combines the generated class with any supplied by the developer. Sorry for the misleading advice previously.

ChristopherHaws commented:
@SteveSandersonMS I updated the test to be an E2E test. I had a hell of a time getting selenium to work because I have v105 of chrome installed but the web driver used by the E2E tests is still v103 and chrome doesn't supply old versions of chrome. I had to lookup the build number of chromium that corresponds to v103 of chrome and use that instead. I remember now why I stopped using selenium and switched to playwright. :)

I think this PR is ready for review now. Thanks!

mkArtakMSFT commented:
Thanks for your PR, @ChristopherHaws.
@SteveSandersonMS could you please review this? Thanks!

mkArtakMSFT commented:
@Nick-Stanton reassigning this to you as @SteveSandersonMS has been busy with other things.

Nick-Stanton commented on src/Components/Web/src/PublicAPI.Shipped.txt:
```diff
@@ -140,6 +140,8 @@ Microsoft.AspNetCore.Components.Forms.ValidationSummary.AdditionalAttributes.get
 Microsoft.AspNetCore.Components.Forms.ValidationSummary.AdditionalAttributes.set -> void
 Microsoft.AspNetCore.Components.Forms.ValidationSummary.Model.get -> object?
 Microsoft.AspNetCore.Components.Forms.ValidationSummary.Model.set -> void
+Microsoft.AspNetCore.Components.Forms.ValidationSummary.CssClass.get -> string?
+Microsoft.AspNetCore.Components.Forms.ValidationSummary.CssClass.set -> void
```
Please move these to the PublicAPI.Unshipped.txt file in the same directory.

ChristopherHaws replied:
Sounds good, I will work on this after work today! :) Thanks!

Nick-Stanton replied:
> Should I update my existing issue with the proposal template or should I make a new issue and link both?

Make a new issue and link anything relevant please 😃

ChristopherHaws replied:
Hi @Nick-Stanton. Sorry it took me a bit to get to this. I have been super busy at work. :)

I moved the new api's to the `PublicAPI.Unshipped.txt` file and created an api proposal for this change here: https://github.com/dotnet/aspnetcore/issues/46461

Nick-Stanton reviewed (changes requested):
Hi @ChristopherHaws, thank you for your contribution and apologies for the delay. I made one small request, but otherwise the changes look good.

Since your PR contains new API, you will need to create an API Proposal to go along with your issue. This is done by opening a new issue using the API Proposal template. More information about the process is [here](https://github.com/dotnet/aspnetcore/blob/main/docs/APIReviewProcess.md). Once the proposal is ready, I will "champion" your changes on your behalf during review. Please let me know if you have any questions.

Nick-Stanton commented:
/azp run

ChristopherHaws committed:
Move changes to unshipped api file

Nick-Stanton commented:
@ChristopherHaws thanks for taking the time to draft your proposal. It looks like the commit that moved your new API caused some new build failures. I imagine CI will be clean once those are fixed.

While investigating these changes deeper to prepare for API review, I had a couple questions:

1. First, I was looking at your example in #43860. If you were to pass `class="validation-errors pt-2"`, wouldn't you get your desired output? I ask because if that scenario works and this feature hasn't been asked for elsewhere, it may be hard to argue for this change. Example:
```html
<ValidationSummary class="validation-errors pt-2" data-test="validation-errors" />
```
2. This implementation has `ValidationSummary` modifying its own CssClass parameter, which you were right to ask about here https://github.com/dotnet/aspnetcore/issues/43860#issuecomment-1244145778. We [usually guide against doing this](https://learn.microsoft.com/aspnet/core/blazor/components/?view=aspnetcore-7.0#overwritten-parameters). Is there a way to achieve this functionality without directly changing `CssClass`?

ChristopherHaws commented:
@Nick-Stanton Hmm, not sure why the build would be failing. Unfortunately I don't have time to look at that this week, I am super busy at work and have a theater show I am running at night.

1. The existing `<ValidationSummary />` component adds `validation-errors`, not the end user.
2. If I remember correctly (it's been almost 6 months since I wrote this code), the CssClass property needs to contain the class names being appended to the component which includes the built-in `validation-errors` css class. 

The only built-in component I found which allows users to set `CssClass` is [NavLink](https://github.com/dotnet/dotnet/blob/febdb074f5ada900f6ef37e35f297b0fced54e8c/src/aspnetcore/src/Components/Web/src/Routing/NavLink.cs#L81)) which also appends it's built-in css class to the property.

Most of the built-in components don't allow you to pass CssClass ([it is get only](https://github.com/dotnet/dotnet/blob/b8bc661a7429baa89070f3bee636b7fbc1309489/src/aspnetcore/src/Components/Web/src/Forms/InputBase.cs#L164)).

I don't know the exact reason that the built-in css class names need to be in this property but I am guessing it is one or many of the following:
- Performance. If you append the string only one time when the parameter is set, you avoid string allocations every time the component gets rendered/re-rendered
- If someone has a ref to the component, they could then get the `CssClass` and it would actually contain all of the css classes, not just the ones the user added
- When writing tests, this would allow you to assert that the css classes are correct without needing to validate it using the DOM

SteveSandersonMS commented:
> The only built-in component I found which allows users to set CssClass is [NavLink](https://github.com/dotnet/dotnet/blob/febdb074f5ada900f6ef37e35f297b0fced54e8c/src/aspnetcore/src/Components/Web/src/Routing/NavLink.cs#L81)) which also appends it's built-in css class to the property.

Ah, checking this more closely, I see that the `CssClass` property you're adding is not actually a `[Parameter]` property so is fine to mutate. (Apologies to @Nick-Stanton - I gave an incorrect suggestion about this.). However I'm not sure it should be settable by subclasses, since their changes would be overwritten every time `OnParametersSet` runs which would be very confusing.

> The existing <ValidationSummary /> component adds validation-errors, not the end user.

That's right, `validation-errors` is the *default* CSS class attribute value. However it's intentional that developers can overwrite this with any other that they want. For example, maybe their styling system doesn't use that CSS class name, so they instead want to use `<ValidationErrors class="my-errors">`. It's intended as a feature that you can replace the class name. It's the same with `ValidationMessage` too.

The change in the PR, if I'm reading correctly, would prevent that and force people to have `validation-errors` whether or not they want it. I know in most cases that won't cause them any problem, but I'm not really sold on the idea that a lot of people are going to be using `class` to represent *additional* class names as opposed to thinking they are setting the entire `class` attribute. Especially given how this would then be inconsistent with `ValidationMessage`.

The reason why we combine classes dynamically in `NavLink` is that `NavLink` adds and removes the `active` class based on navigation state. There has to be *some* dynamic adding/removing of this class since that's the whole point of `NavLink`. However we still don't force any specific hardcoded class name since you can set `ActiveClass` to an arbitrary value.

**In summary** it looks as if we already have the desired functionality, i.e., the ability to customize the CSS class name and add arbitrary extra ones just by specifying `class`. I don't think we'd want to force people to have `validation-errors`. Hope that seems reasonable. If I'm missing something important, please let me know!

ChristopherHaws commented:
@SteveSandersonMS The current behavior of `ValidationSummary` is that you simply cannot override the `class` attribute at all. If you pass a `class`, it is simply ignored. Here are two examples I just tested:
```html
<!-- <ValidationSummary /> -->
<ul class="validation-errors">
    <li class="validation-message">...</li>
</ul>

<!-- <ValidationSummary class="test" data-test="test" /> -->
<ul data-test="test" class="validation-errors">
    <li class="validation-message">...</li>
</ul>
```

I see your point about it being confusing as to whether it replaces or appends the class though. I am coming to blazor after using React for many years where it is pretty common for the classes you pass in to be appended, so I would have assumed that it would be appended. [Here are the docs for FluentUI React](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield#implementation) where you can see that `className` is an `Optional class name that is added to the container of the component.`

I get that this is not React, so I looked at one of the top Blazor Component libraries to see what they do. [MudBlazor](https://www.mudblazor.com/api/field#properties) defines `Class` as `User class names, separated by space.` - The key word here being that this is the "User"'s class names.

SteveSandersonMS commented:
@ChristopherHaws Can you provide a repro using an online REPL?

I just tried it and it works for me: https://blazorrepl.telerik.com/QRumvqaN29a37x4K49

![image](https://user-images.githubusercontent.com/1101362/219325394-313865c1-c19f-4ad9-975e-53993362e9ff.png)

ChristopherHaws commented:
Interesting, I had used [BlazorFiddle](https://blazorfiddle.com/s/6sqte5xq) to try it. I guess they are on an old version of Blazor. It's been a while since I was dealing with this issue in our project (we ended up just wrapping `ValidationSummary` in our own component).
![image](https://user-images.githubusercontent.com/5934505/219567556-7a76f691-d126-4cd0-8ca3-36493ce4ba46.png)

Nick-Stanton commented:
@ChristopherHaws I just looked into it and it looks like they're using a _really_ old version of Blazor.
<img width="520" alt="image" src="https://user-images.githubusercontent.com/25084315/219764349-60a93a7f-e7ee-440b-9bd6-f54430570122.png">
Apologies for not discovering the functionality earlier in review. Do you still have interest in pursuing this PR? It seems like the ask is already met.

mkArtakMSFT commented:
Closing as this is already supported.
